### PR TITLE
[PyOV][DOCS] Python API contribution and developer guide

### PR DESCRIPTION
**Pull request by jiwaszki**
### Details:
 - Added markdown files that describe process of contributing, testing and adding new features (developing) to Python API.
 - New folder that contains a simple example to be used by developers.

**To all reviewers:** this guide takes a little informal approach to the topic, feel free to discuss the form. There are also comments and WIP sections left in first version, please let me hear your feedback about them. Thanks!

### Tickets:
 - 81824, 81830

### CODE REVIEW

File: src/bindings/python/docs/code_examples.md [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+##### Building
+<!-- Link to building instructions -->
+*To be added...*
+
+### Different ways of extending OpenVINO:tm: Python API
+
+###### Before start: Project's naming conventions
+General guide:
+* Snake case is used across the code. That includes modules (`runtime`, `offline_transformations`), function names and arguments/varaibles (`async_infer`, `wait`, `path_to_xml`).
+* Exception to above rule is naming of classes. For example: `Core`, `InferRequest` and `AsyncInferQueue` start each word with a capital letter.
```
Comment (sikorsl1) [inline src/bindings/python/docs/code_examples.md:21]: In the line above you used the term "snake case". I think, that you could use the term "camel case" when describing the style for classes.

File: src/bindings/python/docs/code_examples.md [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+from openvino.helpers.custom_module.custom_helpers import top1_index
+```
+
+Follow it with correct import in `helpers/__init__.py`:
+```python
+from openvino.helpers.custom_module import top1_index
+```
+**Do not forget to include a license on the top of each file!** For demonstrational puroposes (and to keep this markdown file short) it is not included in above snippets.
+
+To see changes taking effect, rebuild project (Cmake's install step should be good enough) and run your solution:
```
Comment (sikorsl1) [inline src/bindings/python/docs/code_examples.md:97]: What about putting here a link to the building tutorial?

File: src/bindings/python/docs/code_examples.md [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+
+<!-- Pure pybind11 solution describes C++ based approach -->
+
+#### Pure pybind11 solution
+Second approach to extend OpenVINO:tm: codebase is utilizing *pybind11* library. It allows to write C++ based code, thus creating so-called Python bindings.
+
+**The example in this section covers the scenario of adding new features to newly created submodule. Extending existing codebase can be done in similar fashion by working on already implemented classes and modules.**
+
+###### Before start: What is pybind11?
+<!-- Do we keep it here? -->
```
Comment (sikorsl1) [inline src/bindings/python/docs/code_examples.md:116]: Yes, I think that this is a good idea to keep it here :)

File: src/bindings/python/docs/code_examples.md [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+
+Add new submodule by writing:
+```cpp
+py::module mymodule = m.def_submodule("mymodule", "My first feature - openvino.runtime.mymodule");
+```
+This is a shorthand way of adding new submodules which can later be used to extend package. The mysterious `m` is actaully the main OpenVINO:tm: module called `pyopenvino` -- it is registered with `PYBIND11_MODULE(pyopenvino, m)` at the top of the "registering-point" file. Later imports from it are done by calling upon `openvino.pyopenvino` package.
+
+Please keep in mind, in most of real-life scenarios, modules and classes are registered in different files. General idea is to create helper function that will hold all of the registered modules, classes and functions. This function needs to be exposed within a separate header file and included in "registering-point". Project's common guideline suggest to use names in the following convention `reg[class|module]_[domain]_[name_of_the_class|module]`. Where optional `[domain]` generally points to parts of the API like graph, frontend or stay empty in case of core runtime. Examples can be found in "registering-point" file `openvino/src/bindings/python/src/pyopenvino/pyopenvino.cpp`.
```
Comment (sikorsl1) [inline src/bindings/python/docs/code_examples.md:133]: What does the sign " | " mean? Is it an alternative? In my opinion splitting `reg[class|module]_[domain]_[name_of_the_class|module]` to the `reg[class]_[domain]_[name_of_the_class] or reg[module]_[domain]_[name_of_the_module]` will be more readable.

File: src/bindings/python/docs/code_examples.md [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+>>> Traceback (most recent call last):
+>>>   File "<stdin>", line 1, in <module>
+>>> TypeError: say_hello(): incompatible function arguments. The following argument types are supported:
+>>>     1. (self: openvino.pyopenvino.mymodule.MyTensor) -> None
+>>>     2. (self: openvino.pyopenvino.mymodule.MyTensor, arg0: str) -> None
+>>> 
+>>> Invoked with: <openvino.pyopenvino.mymodule.MyTensor object at >>> 0x7fdfef5bb4f0>, 777
+```
+Notice that only functions with correct arguments are **not** throwing exceptions. It is helpful to combine this method with your code when binding templated or multi-argument functions. Most of the time (but not always!), using this approach saves a lot of written code, reducing `if-else/switch-case` blocks to minimum and thus making code cleaner and easier to understand.
+<!-- Link do one of our classes? --->
```
Comment (sikorsl1) [inline src/bindings/python/docs/code_examples.md:303]: Link *to ...

File: src/bindings/python/docs/test_examples.md [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+```
+
+Notice that test name is shared between cases. In real-life pull request, all of the functionalities should be tested to ensure the quality of solution. Always focus on general usage and edge-case scenarios. Remember, on the other hand, excessive testing is also not advised as it may result in duplicate test cases and taking a toll on validation pipelines. Good "rule-of-thumb"s while adding tests to the project:
+* Don't test built-in capabilities of a given language.
+* Common functions can be tested together.
+* Create test cases with few standard scenarios and cover all known edge-cases.  
+* Hardcode desired results...
+* ... or create reference values during runtime. Always use a good, thrust-worthy library for that!
+* Re-use common parts of the code (like multiple lines that create helper object) and move them out to make tests easier to read.
+* My personal advice: take a little break and start writing tests with fresh mind, try to take a user perspective.
```
Comment (sikorsl1) [inline src/bindings/python/docs/test_examples.md:114]: I would recommend not to use the term "My personal advice" unless this guide is written from first person perspective.

**Review by sikorsl1 — APPROVED** [review]
Please treat my comments as a proposition not as a requirement :)

File: src/bindings/python/docs/code_examples.md [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+
+###### Before start: Project's naming conventions
+General guide:
+* Snake case is used across the code. That includes modules (`runtime`, `offline_transformations`), function names and arguments/varaibles (`async_infer`, `wait`, `path_to_xml`).
+* Exception to above rule is naming of classes. For example: `Core`, `InferRequest` and `AsyncInferQueue` start each word with a capital letter.
+* If bindings (term explained later in section "Pure pybind11 solution" <!-- Add link to section? https://stackoverflow.com/questions/2822089/how-to-link-to-part-of-the-same-document-in-markdown -->) are created to already existing C++ code, try to get close to their C++ counterpart with names and placement (example: C++'s `ov::InferRequest` and Python's `openvino.runtime.InferRequest`). If aligement is not possible, try to describe your class/function/module in the best way. One of the examples is pair of `openvino.runtime.ConstOutput/openvino.runtime.Output` which relates to `ov::Output<const ov::Node>/ov::Output<ov::Node>`, this naming points out to functional difference between both classes - first is immutable version and second is mutable.
```
Comment (tomdol) [inline src/bindings/python/docs/code_examples.md:22]:
```suggestion
* If bindings (term explained later in section "Pure pybind11 solution" <!-- Add link to section? https://stackoverflow.com/questions/2822089/how-to-link-to-part-of-the-same-document-in-markdown -->) are created to expose existing C++ code, try to get close to their C++ counterpart with names and placement (example: C++'s `ov::InferRequest` and Python's `openvino.runtime.InferRequest`). If alignment is not possible, try to describe your class/function/module in the best way. One of the examples is pair of `openvino.runtime.ConstOutput/openvino.runtime.Output` which relates to `ov::Output<const ov::Node>/ov::Output<ov::Node>`, this naming points out the functional difference between both classes - the first is immutable version and the second is mutable.
```

File: src/bindings/python/docs/code_examples.md [hunk 8]
```diff
@@ -0,0 +1,413 @@
...
+
+<!-- Pure Python solution describes Python based approach -->
+#### Pure Python solution
+One of the simplest way to extend existing codebase is by writing it in pure Python.
+
+###### Before start: Layout of the project
+How is OpenVINO:tm: packaging working? It is strictly connected to the layout of the Python API itself and reused in different supporting packages like tools and extensions. Main namespace `openvino` provides unified place that connects all packages together during importing, **this is the required part**. However, it is up to the developer how to organize the rest of the package. There are also other common namespaces which follows same rules:
```
Comment (tomdol) [inline src/bindings/python/docs/code_examples.md:29]:
```suggestion
How does OpenVINO:tm: packaging work? It is strictly connected to the layout of the Python API itself and reused in different supporting packages like tools and extensions. The main namespace `openvino` provides unified place that connects all packages together during importing, **this is the required part**. However, it is up to the developer how to organize the rest of the package. There are also other common namespaces which follow the same rules:
```

File: src/bindings/python/docs/code_examples.md (hunk 2 above)
Comment (tomdol) [inline src/bindings/python/docs/code_examples.md:97]:
```suggestion
To see changes taking effect rebuild the project (CMake's install step should be good enough) and run your solution:
```

File: src/bindings/python/docs/code_examples.md [hunk 9]
```diff
@@ -0,0 +1,413 @@
...
+*To be added...*
+
+##### Building
+<!-- Link to building instructions -->
+*To be added...*
+
+### Different ways of extending OpenVINO:tm: Python API
+
+###### Before start: Project's naming conventions
+General guide:
```
Comment (praasz) [inline src/bindings/python/docs/code_examples.md:19]: I'm not sure if that true but looks like naming convention is like [PEP8 naming convention](https://peps.python.org/pep-0008/#naming-conventions)?

Maybe mention it here and write what is a different from this standard?

In PEP there is also mention about styles:
* UPPERCASE
* UPPER_CASE_WITH_UNDERSCORES
etc.

In what cases we can used here or maybe is not allowed at all.

File: src/bindings/python/docs/contributing.md [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    openvino/src/bindings/python/docs/code_example.md
+
+##### Always test out our code! Don't forget about it before pushing and triggering CIs.
+
+Please refer to Test Guide available here:
+
+    openvino/src/bindings/python/docs/test_examples.md
+
+Moreover, the project utilizes *flake8* and *mypy* packages to run codestyle checks. Additionally OpenVINO:tm: uses the custom configuration file to exclude some strict rules. To run codestyle checks, navigate to main Python API folder first and use following commands:
+```
```
Comment (praasz) [inline src/bindings/python/docs/contributing.md:33]: Detail: Add shell/bash here to enable some code block colorization?

e.g.:
```shell
# Comment
cd .../openvino/src/bindings/python/
flake8 src/openvino/ --config=setup.cfg
```

File: src/bindings/python/docs/examples/openvino/mymodule/myclass.py [hunk 11]
```diff
@@ -0,0 +1,11 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2018-2022 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+class MyClass():
+
```
Comment (praasz) [inline src/bindings/python/docs/examples/openvino/mymodule/myclass.py:6]: Consider provide example of docstrings for class/methods and what format shall be used e.g.: epydoc, google, reST etc.

File: src/bindings/python/docs/test_examples.md (hunk 6 above)
Comment (jiwaszki) [reply src/bindings/python/docs/test_examples.md:114]: Got it!

File: src/bindings/python/docs/code_examples.md (hunk 9 above)
Comment (jiwaszki) [reply src/bindings/python/docs/code_examples.md:19]: I forgot that PEP8 actually mentioned it, thank you for a pointer!

File: src/bindings/python/docs/code_examples.md (hunk 5 above)
Comment (jiwaszki) [reply src/bindings/python/docs/code_examples.md:303]: Link do-be-do-be-doo

File: src/bindings/python/docs/code_examples.md (hunk 2 above)
Comment (jiwaszki) [reply src/bindings/python/docs/code_examples.md:97]: Will do when `BUILDING.md` will be updated, good idea!

File: src/bindings/python/docs/code_examples.md (hunk 4 above)
Comment (jiwaszki) [reply src/bindings/python/docs/code_examples.md:133]: Good idea!

File: src/bindings/python/docs/examples/openvino/mymodule/myclass.py (hunk 11 above)
Comment (jiwaszki) [reply src/bindings/python/docs/examples/openvino/mymodule/myclass.py:6]: Will do!

**Review by praasz — APPROVED** [review]

File: src/bindings/python/docs/code_examples.md [hunk 12]
```diff
@@ -0,0 +1,414 @@
...
+    ...
+    runtime/
+    test_utils/
+    __init__.py
+    utils.py
+```
+
+Let's add in `custom_module/custom_helpers.py`:
+```python
+def top1_index(results: list) -> int:
```
Comment (p-wysocki) [inline src/bindings/python/docs/code_examples.md:82]: It could be renamed to `argmax` to make the functions purpose obvious from the start, without looking at its body to make sure. Technically argmax can return a set of points instead of a single index (so it may not be a precise name), so it's up to your decision.

File: src/bindings/python/docs/code_examples.md [hunk 13]
```diff
@@ -0,0 +1,414 @@
...
+```
+
+*Tip: To add a function on the module's level as a "free function", simply define it on the module's object:*
+```cpp
+mymodule.def("get_smile", []() {
+    py::print(":)");
+});
+```
+
+**Don't forget to rebuild the project** and test it out:
```
Comment (p-wysocki) [inline src/bindings/python/docs/code_examples.md:223]: Would being explicit about rebuilding AND reinstalling the project (`make install` is required for Python changes to be applied) be beneficial? Without `make install` the contributor could get confused, not noticing new behavior in Python.

File: src/bindings/python/docs/code_examples.md (hunk 12 above)
Comment (jiwaszki) [reply src/bindings/python/docs/code_examples.md:82]: I didn't consider `argmax` as a name since it would conflict with numpy function. I would like to keep it as-is.

File: src/bindings/python/docs/code_examples.md (hunk 13 above)
Comment (jiwaszki) [reply src/bindings/python/docs/code_examples.md:223]: Yeah... actually it will depend on `BUILDING.md` instructions, so I am going to keep this comment unresolved and get back to it as soon as Building Guide PR starts.

**Review by kblaszczak-intel — COMMENTED** [review]
That is a very interesting set of articles. It was a good read :)
When it goes to corrections, they are almost exclusively minor grammar tweaks, mostly regarding the use of articles. That is why I put them as suggestions, so that they would be easy to accept/reject.
A couple of rewordings have been suggested, but they are quite safe. One is provided with an additional comment/explanation.

File: src/bindings/python/docs/code_examples.md [hunk 14]
```diff
@@ -0,0 +1,414 @@
...
+
+a.say_hello([1,2,3])
+>>> Traceback (most recent call last):
+>>>   File "<stdin>", line 1, in <module>
+>>>   File ".../openvino/runtime/mymodule_ext.py", line 20, in say_hello
+>>>     raise TypeError("Unsupported argument type!")
+>>> TypeError: Unsupported argument type!
+```
+
+Great! Now the class has gone a full cricle, from C++ to Python (and again to Python). Such aliasing is a common technique in the project and gives a lot of power to the developer. With easy to understand code, `say_hello` function is now able to dispatch arguments based on their type and apply necessary preprocessing to feed data into the function. However, this could be done in "few more lines of C++" as well, where is the tricky part? The difficult feat of returning different types based on same argument type is achived (look at the dispatching of integer arguments).
```
[the remark this answers is not in the document]
Comment (jiwaszki) [reply src/bindings/python/docs/code_examples.md:405]: Good catch! I doubted it myself at some point... 😄Let's keep your suggestion, I like it even more.

**Review by tomdol — APPROVED** [review]
